source/agora/common/BlockStorage.d

```diff
@@ -159,35 +298,346 @@ public class BlockStorage
+            const size_t block_size = this.readSizeT(MFILE_HEAD_SIZE + x0);
+
+            assert(x0 < this.data_size);
+
+            const size_t x2 = x0 + MFILE_HEAD_SIZE + size_t.sizeof;
+            const size_t x3 = x2 + block_size;
+
+            block = deserialize!Block(this.read(x2, x3));
+
+            return true;
+        }
         else
-            to = this.block_positions[next_height];
+        {
+            return false;
+        }
+    }
 
-        assert(from < file.length);
-        assert(from < to);
+    /***************************************************************************
 
-        block = deserialize!Block(cast(ubyte[])this.file[from .. to]);
+        Read block from the file.
 
-        return true;
+        Params:
+            block = `Block` to read
+            hash = `Hash` of `Block`
+
+        Returns:
+            Returns true if success, otherwise returns false.
+
+    ***************************************************************************/
+
+    public bool readBlock (ref Block block, Hash hash) @safe
+    {
+        ubyte[HASH_SIZE] hash_bytes = hash[0..HASH_SIZE];
+
+        auto finds
+            = this.hash_idx[].find!((a, b) => a.hash == b)(hash_bytes);
+
+        if (!finds.empty)
+        {
+            const size_t position = finds.front.position;
+
+            this.map(position / MFILE_MAP_SIZE);
+
+            const size_t x0 = position % MFILE_MAP_SIZE;
+            const size_t block_size = this.readSizeT(MFILE_HEAD_SIZE + x0);
+
+            assert(x0 < this.data_size);
+
+            const size_t x2 = x0 + MFILE_HEAD_SIZE + size_t.sizeof;
+            const size_t x3 = x2 + block_size;
+
+            block = deserialize!Block(this.read(x2, x3));
+
+            return true;
+        }
+        else
+        {
+            return false;
+        }
     }
+
+    /***************************************************************************
+
+        Write data length to file header
+
+        Params:
+            length = The size of data
+
+    ***************************************************************************/
+
+    private void writeDataLength (size_t length) @safe
+    {
+        this.writeSizeT(0, length);
+    }
+
+    /***************************************************************************
+
+        Read data length to file header
+
+        Returns:
+            The size of data in file header
+
+    ***************************************************************************/
+
+    private size_t readDataLength () @safe
+    {
+        return this.readSizeT(0);
+    }
+
+    /***************************************************************************
+
+        Read type of `size_t` data
+
+        Params:
+            pos = position of memory mapped file
+            value = type of `size_t`
+
+    ***************************************************************************/
+
+    private void writeSizeT (size_t pos, size_t value) @trusted
```

Comment:
W.r.t. `readDataLength`, `writeDataLength`, `writeSizeT`, `readSizeT`: This is what the serialization / deserialization module is for.